### ISSUE TITLE
Nested changesets

### ISSUE THREAD

**Issue by gjaldon** [elixir-lang/ecto#618]
In cases where we'd want to create a `User` and require that it has an associated `Address`, it would be great to be able to call `Repo.insert(changeset)` and have Ecto create both a `User` and its associated `Address`. We could do this by having the changeset for `Address` stored in the changeset for `User` when we call `cast/4` on `User` with the `:address` association as one of its required 'fields'. 

For example, in the User model we'll set the required fields like:

```elixir
defmodule MyApp.User do
  schema "users" do
    field :name, :string
    field :email, :string
    has_one :address, MyApp.Address
  end

 @required_fields ~w(name email address)

  def changeset(model, params \\ :empty) do
    model
    |> cast(params, @required_fields, [])
  end
end
```

When passing an association as one of the required or optional fields in `cast/4`, we call the assocation's `changeset/2`. So the `cast/4` above will call `MyApp.Address.changeset/2`, expecting a key `"address"` in the params. We call `MyApp.Address.changeset/2` so that it calls `cast/4` and the validations for `MyApp.Address` (`changeset/2` will be a required function in Ecto Model instead of just a convention).

In summary, I'm proposing the ff:
- passing an association as a required or optional field to `cast/4` will call that association's `changeset/2` function
- `changeset/2` be a required function in all `Ecto.Model`s
- the association changeset will be stored in the primary model's changeset. Not sure what to name the field yet. Maybe `:nested`?
- errors in a nested changeset will bubble up to the parent changeset and set `valid?` to `false`
- Make `Repo.update/1` and `Repo.insert/1` recognize nested changesets and insert/update the nested changesets along with the primary one

What do you guys think?

**Comment by whatyouhide** [comment]
I'm a little scared by the second point, making `changeset/2` required in all models; I think it may limit the generality of an `Ecto.Model`. Couldn't we just raise an error on missing `changeset/2` (maybe with an option somewhere to specify a different function to call? Provided it could be implemented in a simple way).

**Comment by josevalim** [comment]
This is roughly what is intended for #357. There are many details I am not sure yet and we have more pressing concerns to handle before 1.0. So the plan is to tackle this after 1.0 is out. :)

**Comment by gjaldon** [comment]
Ok will follow that issue then :)
